include/circt/Dialect/Verif/VerifOps.td

```diff
@@ -346,6 +349,105 @@ def FormalOp : VerifOp<"formal", [
+    clock, and then drops to 0 for the remainder of the simulation.
+
+    ### Outputs
+
+    The body must have a `verif.yield` terminator op with exactly two operands:
+
+    The first operand is a "done" signal of type `i1` which indicates the end of
+    the simulation. The simulation stops when the done signal is 1 during a
+    0-to-1 transition of the clock.
+
+    The second operand is an "exit code" signal which may be of any integer
+    type. It indicates success of a test as 0, or failure as any non-zero
+    integer. Only the zero versus non-zero value distinction is relevant to
+    indicate simulation success and must be preserved. The exact non-zero value
+    is for informational purposes only and may not be reproduced accurately by
+    simulators. Simulators may in the worst case map all non-zero values to 1,
+    reducing the exit code to a single bit indicating failure or success.
+
+    ### Schedule
+
+    The clock and init values adhere to the following schedule during
+    simulation:
+
+    | Time      | Clock | Init  | Comment
+    |-----------|-------|-------| ----
+    | t0 (>=0s) | undef | undef | Clock and init may initially be undefined.
+    | t1 (>=t0) | 0     | 1     | Initialization code (e.g., `seq.initial`, Verilog `initial` procedures) may run before or after clock and init change to their initial value.
+    | t2 (>t1)  | 1     | 1     | Single rising clock edge occurs while init is high.
+    | t3 (>t2)  | 0     | 1     | Init may stay during the falling clock edge.
+    | t4 (>=t3) | 0     | 0     | Init goes to 0 before second rising clock edge.
+    | t5 (>t4)  | 1     | 0     | Clock toggles continue indefinitely.
+    | t6 (>t5)  | 0     | 0     |
+
+    Simulation termination occurs when the done signal is 1 during a 0-to-1
+    transition of the clock. The exit code is sampled at the same time as the
+    done signal. No additional clock toggles occur once done has been sampled as
+    1.
+
+    ### Example
+    ```
+    verif.simulation @AdderTest {myParam = 42, myTag = "hello"} {
+    ^bb0(%clock: !seq.clock, %init: i1):
+      // Count the first 9001 simulation cycles.
+      %c0_i19 = hw.constant 0 : i19
+      %c1_i19 = hw.constant 1 : i19
+      %c9001_i19 = hw.constant 9001 : i19
+      %count = seq.compreg %0, %clock reset %init, %c0_i19 : i19
+      %done = comb.icmp eq %count, %c9001_i19 : i19
+      %0 = comb.add %count, %c1_i19 : i19
+
+      // Generate inputs to the adder.
+      %1, %2 = func.call @generateAdderInputs(%count) : (i19) -> (i42, i42)
+      %3 = hw.instance "dut" @Adder(a: %1: i42, b: %2: i42) -> (c: i42)
+
+      // Check results and track failures.
+      %4 = comb.add %1, %2 : i42
+      %5 = comb.icmp ne %3, %4 : i42
+      %false = hw.constant false
+      %failure = seq.compreg %6, %clock reset %init, %false : i1
+      %6 = comb.or %failure, %5 : i1
+
+      verif.yield %done, %failure : i1, i1
+    }
```

Comment:
Is the body supposed to be a hardware implementation of a testbench or are you allowed to do any kind of software stuff? I guess not, because it's a graph region.

Reply:
It's basically the same as a `hw.module` or `verif.formal` op. Any software specifics would have to be packaged up in an op that can describe how a SSACFG region can be executed and can update some form of output values it produces, maybe something like `sim.execute_region` with multiple `sim.yield` resume points.